### ISSUE TITLE
[JENKINS-17774] Testing plugin in Windows fails with requested operation can't be performed on file with user-mapped section open

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#6594]
<p>Running tests of <a href="https://wiki.jenkins-ci.org/display/JENKINS/Extensible+Choice+Parameter+plugin" class="external-link" target="_blank" rel="nofollow noopener">extensible-choice-parameter plugin</a> in Windows often fails.<br/>
This failure is caused by 500 error in Jenkins (or Jetty).<br/>
500 error is caused by java.io.FileNotFoundException with an error message "requested operation can't be performed on file with user-mapped section open".<br/>
(I attached a sunfire-report file generated when this problem happened, but the error message is output in Japanese.)</p>

<ul>
	<li>This probably happens only in Windows (maybe Windows Vista or later)</li>
	<li>FileNotFoundException happens when copying(overwriting) jenkins-for-test to jetty's webapps directory.</li>
	<li>This seems happen only after running hpi:run.
	<ul>
		<li>It probably depends on Jetty's working directory. Before running hpi:run, Jetty's working directory is %TEMP%\Jetty_0_0_0_0_0_jenkins.for.test__<em>.bj8wp3. After running hpi:run, Jetty seems run in (project directory)\work\Jetty_0_0_0_0_0_jenkins.for.test</em>__.bj8wp3 .</li>
		<li>I have no idea why it affects.</li>
	</ul>
	</li>
</ul>

<p>I can reproduce this in <a href="https://github.com/ikedam/extensible-choice-parameter" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/ikedam/extensible-choice-parameter</a>, tag reproduce_23_user-mapped_section.<br/>
And I have fixed it in branch feature/23_JettyOnWindows, by setting useFileMappedBuffer to false in Jetty webapp context (SEE <a href="http://docs.codehaus.org/display/JETTY/Files+locked+on+Windows" class="external-link" target="_blank" rel="nofollow noopener">http://docs.codehaus.org/display/JETTY/Files+locked+on+Windows</a>).</p>

<p>I think this is related to <a href="https://issues.jenkins.io/browse/JENKINS-12647" title="Jetty configuration causes locked files when running tests on Windows, which causes test failures" class="issue-link" data-issue-key="JENKINS-12647"><del>JENKINS-12647</del></a> (I have not succeeded reproducing problems reported in this issue).<br/>
As long as I tested, the problem in extensible-choice-parameter is fixed in a way of followings:</p>

<ul>
	<li>setting useFileMappedBuffer to false, setting setCopyWebDir to true</li>
	<li>setting useFileMappedBuffer to false, setting setCopyWebDir to false</li>
	<li>setting setCopyWebDir to false (disabling <a href="https://issues.jenkins.io/browse/JENKINS-12647" title="Jetty configuration causes locked files when running tests on Windows, which causes test failures" class="issue-link" data-issue-key="JENKINS-12647"><del>JENKINS-12647</del></a>)</li>
</ul>

**Attachments:**

- [TEST-jp.ikedam.jenkins.plugins.extensible_choice_parameter.TextareaChoiceListProviderJenkinsTest.xml](https://issues.jenkins.io/secure/attachment/23489/TEST-jp.ikedam.jenkins.plugins.extensible_choice_parameter.TextareaChoiceListProviderJenkinsTest.xml)

---
<details><summary><i>Originally reported by <a title="ikedam" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ikedam">ikedam</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-17774" target="_blank">Testing plugin in Windows fails with requested operation can't be performed on file with user-mapped section open</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="ikedam" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ikedam">ikedam</a>
<li><b>reported by</b>: ikedam
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2013-06-05T12:55:17+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ikedam">ikedam</a>:</i>
<p>Test failures in Email-ext 2.18 can be resolved with <a href="https://github.com/jenkinsci/jenkins/commit/24e99923e3c593073e704a4e8ebf70d5ed11f313" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/commit/24e99923e3c593073e704a4e8ebf70d5ed11f313</a>.</p>

<p>I conclude that <b><a href="https://issues.jenkins.io/browse/JENKINS-12647" title="Jetty configuration causes locked files when running tests on Windows, which causes test failures" class="issue-link" data-issue-key="JENKINS-12647"><del>JENKINS-12647</del></a> DO NOTHING AT ALL</b>.<br/>
I try to create a patch.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ikedam">ikedam</a>:</i>
<p>Sent a pull request <a href="https://github.com/jenkinsci/jenkins/pull/773" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/773</a> .<br/>
jglick pointed followings:</p>
<ul>
	<li>Extracting createWebAppContext makes the interface of HudsonTestCase depend on Jetty.</li>
	<li>A same change should be applied to JenkinsRule.</li>
</ul>


<p>I found that JenkinsRule does not have this problem(<a href="https://issues.jenkins.io/browse/JENKINS-17774" title="Testing plugin in Windows fails with requested operation can&#39;t be performed on file with user-mapped section open" class="issue-link" data-issue-key="JENKINS-17774"><del>JENKINS-17774</del></a>).<br/>
So this problem can be avoided by using a rule annotation with JenkinsRule, instead of deriving from HudsonTestCase.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ikedam">ikedam</a>:</i>
<p>I verified followings:</p>

<ul>
	<li>With JenkinsRule, <a href="https://issues.jenkins.io/browse/JENKINS-17774" title="Testing plugin in Windows fails with requested operation can&#39;t be performed on file with user-mapped section open" class="issue-link" data-issue-key="JENKINS-17774"><del>JENKINS-17774</del></a> cannot be reproduced: <a href="https://github.com/ikedam/reproduce-jenkins-17774/commit/54ffef8b12cc74c33257baa40b94199a6182be37" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/ikedam/reproduce-jenkins-17774/commit/54ffef8b12cc74c33257baa40b94199a6182be37</a></li>
	<li>By swiching from HudsonTestCase to JenkinsRule, existing plugin gets free from <a href="https://issues.jenkins.io/browse/JENKINS-17774" title="Testing plugin in Windows fails with requested operation can&#39;t be performed on file with user-mapped section open" class="issue-link" data-issue-key="JENKINS-17774"><del>JENKINS-17774</del></a>: <a href="https://github.com/ikedam/extensible-choice-parameter/commit/d06522a9fb32a754eb0e9af90e82ddd0b52f0c8b" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/ikedam/extensible-choice-parameter/commit/d06522a9fb32a754eb0e9af90e82ddd0b52f0c8b</a></li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=ikedam">ikedam</a>:</i>
<p>Created a new pull request just reverting 12647.<br/>
<a href="https://github.com/jenkinsci/jenkins/pull/779" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/779</a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: ikedam<br/>
Path:<br/>
 test/src/main/java/org/jvnet/hudson/test/HudsonTestCase.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/68db55528a560b816fbd58689b75d07210592b03" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/68db55528a560b816fbd58689b75d07210592b03</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-17774&#93;</span> Make not to copy a webapp directory every time for each tests.<br/>
Copying every time causes overwriting files in the webapp directory, and often causes errors in Windows.<br/>
This reverts <a href="https://issues.jenkins.io/browse/JENKINS-12647" title="Jetty configuration causes locked files when running tests on Windows, which causes test failures" class="issue-link" data-issue-key="JENKINS-12647"><del>JENKINS-12647</del></a>(commit 7ae303f88191a540491a21484316d63dff775e21).</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: Jesse Glick<br/>
Path:<br/>
 test/src/main/java/org/jvnet/hudson/test/HudsonTestCase.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/e3fee0aa36d4afe3d7e17e17c766c4230758c41d" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/e3fee0aa36d4afe3d7e17e17c766c4230758c41d</a><br/>
Log:<br/>
  Merge pull request #779 from ikedam/feature/<a href="https://issues.jenkins.io/browse/JENKINS-17774" title="Testing plugin in Windows fails with requested operation can&#39;t be performed on file with user-mapped section open" class="issue-link" data-issue-key="JENKINS-17774"><del>JENKINS-17774</del></a>-2</p>

<p><span class="error">&#91;FIXED JENKINS-17774&#93;</span> Plugin tests often fails in Windows (just reverting <a href="https://issues.jenkins.io/browse/JENKINS-12647" title="Jetty configuration causes locked files when running tests on Windows, which causes test failures" class="issue-link" data-issue-key="JENKINS-12647"><del>JENKINS-12647</del></a>).</p>


<p>Compare: <a href="https://github.com/jenkinsci/jenkins/compare/63ecb8269aaa...e3fee0aa36d4" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/compare/63ecb8269aaa...e3fee0aa36d4</a></p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=scm_issue_link">scm_issue_link</a>:</i>
<p>Code changed in jenkins<br/>
User: ikedam<br/>
Path:<br/>
 test/src/main/java/org/jvnet/hudson/test/HudsonTestCase.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins-test-harness/a9eb82b768ed88cf09be66dc4041dcc0c34777c1" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins-test-harness/a9eb82b768ed88cf09be66dc4041dcc0c34777c1</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-17774&#93;</span> Make not to copy a webapp directory every time for each tests.<br/>
Copying every time causes overwriting files in the webapp directory, and often causes errors in Windows.<br/>
This reverts <a href="https://issues.jenkins.io/browse/JENKINS-12647" title="Jetty configuration causes locked files when running tests on Windows, which causes test failures" class="issue-link" data-issue-key="JENKINS-12647"><del>JENKINS-12647</del></a>(commit 7ae303f88191a540491a21484316d63dff775e21).</p>

<p>Originally-Committed-As: 68db55528a560b816fbd58689b75d07210592b03</p>
